parcels/scripts/utils.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import xarray as xr
+from os import path
+from netCDF4 import Dataset
+
+
+def compute_curvilinear_rotation_angles(mesh_filename, rotation_angles_filename, variables, dimensions):
+
+    if path.isfile(rotation_angles_filename) and path.getmtime(rotation_angles_filename) > path.getmtime(mesh_filename):
+        return
+
+    dataset = xr.open_dataset(mesh_filename)
+    lonU = np.squeeze(getattr(dataset, dimensions['U']['lon']).values)
+    latU = np.squeeze(getattr(dataset, dimensions['U']['lat']).values)
+    lonV = np.squeeze(getattr(dataset, dimensions['V']['lon']).values)
+    latV = np.squeeze(getattr(dataset, dimensions['V']['lat']).values)
+    lonF = np.squeeze(getattr(dataset, dimensions['F']['lon']).values)
+    latF = np.squeeze(getattr(dataset, dimensions['F']['lat']).values)
+    dataset.close()
+
+    rad = np.pi / 180.
+    rpi = np.pi
+
+    zxnpu = - 2. * np.cos(rad*lonU[1:, 1:]) * np.tan(rpi/4. - rad*latU[1:, 1:]/2.)
+    zynpu = - 2. * np.sin(rad*lonU[1:, 1:]) * np.tan(rpi/4. - rad*latU[1:, 1:]/2.)
+    znnpu = zxnpu*zxnpu + zynpu*zynpu
+
+    zxnpv = - 2. * np.cos(rad*lonV[1:, 1:]) * np.tan(rpi/4. - rad*latV[1:, 1:]/2.)
+    zynpv = - 2. * np.sin(rad*lonV[1:, 1:]) * np.tan(rpi/4. - rad*latV[1:, 1:]/2.)
+    znnpv = zxnpv*zxnpv + zynpv*zynpv
+
+    zxffu = 2. * np.cos(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.cos(rad*lonF[:-1, 1:]) * np.tan(rpi/4. - rad*latF[:-1, 1:]/2.)
+    zyffu = 2. * np.sin(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.sin(rad*lonF[:-1, 1:]) * np.tan(rpi/4. - rad*latF[:-1, 1:]/2.)
+    znffu = np.sqrt(znnpu * (zxffu*zxffu + zyffu*zyffu))
+    znffu = np.maximum(znffu, 1.e-14)
+
+    zxffv = 2. * np.cos(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.cos(rad*lonF[1:, :-1]) * np.tan(rpi/4. - rad*latF[1:, :-1]/2.)
+    zyffv = 2. * np.sin(rad*lonF[1:, 1:]) * np.tan(rpi/4. - rad*latF[1:, 1:]/2.) \
+        - 2. * np.sin(rad*lonF[1:, :-1]) * np.tan(rpi/4. - rad*latF[1:, :-1]/2.)
+    znffv = np.sqrt(znnpv * (zxffv*zxffv + zyffv*zyffv))
+    znffv = np.maximum(znffv, 1.e-14)
```

Comment:
seems a lot of repetition. Use functions?

Reply:
It's a translation of the Fortran implementation. The link to the Fortran file is given in comments